### DISCUSSION TITLE
Using Tok2Vec listener with TextCat

### DISCUSSION THREAD

**Discussion by Economax** [explosion/spaCy#7811] [Help: Coding & Implementations]
I wanted to explore how a TextCat pipeline would perform with a Tok2Vec listener implementation rather than the standard sub-layer implementation, but that has proven difficult. Am I doing something wrong, or is the TextCat pipeline simply not designed to listen to a Tok2Vec from earlier in the pipeline?

I'm using the "nb_core_news_lg" language model, and then adding a standard "textcat" pipeline at the end, modifying the textcat pipeline config to use a Tok2Vec listener rather than building its own.

The error message is the following:

ValueError: [E954] The Tok2Vec listener did not receive any valid input from an upstream component.

Which happens during the Training Pipeline. This error only occurs with the following config file specification of the Tok2Vec layer in the TextCat:

`[components.textcat.model.tok2vec]`
`@architectures = "spacy.Tok2VecListener.v1"`
`width = 96`
`upstream = "*"`

**Comment by svlandeg** [comment]
The `textcat` should be able to listen to a `Tok2Vec` layer. Can you paste the full config? From the error message, is looks as if the original `tok2vec` layer didn't run, but with the full config file it will be easier to debug this.

**Comment by Economax** [comment]
Full config:

```
[paths]
train = "corpus/nb-core-news/train.spacy"
dev = "corpus/nb-core-news/dev.spacy"
vectors = "nb_core_news_lg"
raw = null
init_tok2vec = null
vocab_data = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "nb"
pipeline = ["tok2vec","morphologizer","parser","senter","ner","attribute_ruler","lemmatizer","textcat"]
before_creation = null
after_creation = null
after_pipeline_creation = null
batch_size = 256
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.attribute_ruler]
source = "nb_core_news_lg"


#[components.textcat.model.tok2vec]
#@architectures = "spacy.Tok2Vec.v2"

#[components.textcat.model.tok2vec.embed]
#@architectures = "spacy.MultiHashEmbed.v1"
#width = 64
#rows = [2000,2000,1000,1000,1000,1000]
#attrs = ["ORTH","LOWER","PREFIX","SUFFIX","SHAPE","ID"]
#include_static_vectors = false

#[components.textcat.model.tok2vec.encode]
#@architectures = "spacy.MaxoutWindowEncoder.v2"
#width = 64
#window_size = 1
#maxout_pieces = 3
#depth = 4


[components.lemmatizer]
source = "nb_core_news_lg"

[components.morphologizer]
source = "nb_core_news_lg"

[components.ner]
source = "nb_core_news_lg"

[components.parser]
source = "nb_core_news_lg"

[components.senter]
source = "nb_core_news_lg"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v1"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.tok2vec.model.encode:width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v1"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[components.textcat]
factory = "textcat"
threshold = 0.5

[components.textcat.model]
@architectures = "spacy.TextCatEnsemble.v2"
nO = null

[components.textcat.model.linear_model]
@architectures = "spacy.TextCatBOW.v1"
exclusive_classes = true
ngram_size = 3
no_output_layer = false
nO = null

[components.textcat.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode:width}
upstream = "*"

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[training]
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = ["tok2vec","morphologizer","parser","senter","ner","attribute_ruler","lemmatizer"]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
cats_score_desc = null
cats_micro_p = null
cats_micro_r = null
cats_micro_f = null
cats_macro_p = null
cats_macro_r = null
cats_macro_f = null
cats_macro_auc = null
cats_f_per_type = null
cats_macro_auc_per_type = null
cats_score = 1.0

[pretraining]

[initialize]
vocab_data = ${paths.vocab_data}
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
before_init = null
after_init = null

[initialize.components]

[initialize.components.morphologizer]

[initialize.components.morphologizer.labels]
@readers = "spacy.read_labels.v1"
path = "corpus/labels/morphologizer.json"
require = false

[initialize.components.ner]

[initialize.components.ner.labels]
@readers = "spacy.read_labels.v1"
path = "corpus/labels/ner.json"
require = false

[initialize.components.parser]

[initialize.components.parser.labels]
@readers = "spacy.read_labels.v1"
path = "corpus/labels/parser.json"
require = false

[initialize.lookups]
@misc = "spacy.LookupsDataLoader.v1"
lang = ${nlp.lang}
tables = []

[initialize.tokenizer]

```

**Reply by adrianeboyd** [reply]
Hi, there are two problems with the config:

* your pipeline is missing the `tok2vec` component from `nl_core_news_lg`, which is required by the `morphologizer` and `parser`, so you need to `source` it from `nl_core_web_lg`, too
* if the `tok2vec` component is frozen, then it doesn't run during training and the downstream component fails (this is the direct error you're seeing)

For the `tok2vec` for `textcat`, I think there are a few options:

1. have it listen to a separate `tok2vec` component with a custom name like `textcat_tok2vec` and modify the `upstream` settings for all components to point the right listeners
2. have an internal `tok2vec` model in the `textcat` component 

There's no real advantage of 1 over 2 since the `textcat` component is the only component using it, so 2 is easier to configure and avoids the complicated naming/listener situation. I'm not sure you can even make the required changes to `upstream` correctly for sourced components from a config at this point, either.

If you configure your `textcat` component to listen to the sourced `tok2vec` from `nl_core_news_lg` and it's not frozen, then training `textcat` will degrade the performance for all the other components that listen to it (`morphologizer` and `parser`), so this isn't a good idea.

It will be possible to have a component listen to a frozen `tok2vec` with a new training option in v3.1+ (not yet released), but I just tried it out with one of the textcat demos + `tok2vec` from `en_core_web_sm` and the performance dropped a lot, which isn't surprising since the `tok2vec` is a core part of the model. There may be situations where this could work, but it's not a good general recommendation.

**Reply by Economax** [reply]
1. Pipeline missing: Thanks, I'll get to fixing that and re-run the config!
2. Tok2Vec frozen: I was under the impression that active components during training are also trained. Is this not the case?

On 2: I want to use the pre-trained Tok2Vec from nb_core_news_lg, and I don't want to update its models during the training stage. I do want to update the TextCat model during training. I was under the impression that freezing components allowed the components to be run during training, but not be updated. Furthermore, I thought that disabled pipes were not ran during training. Otherwise, why use two different settings for the same functionality?

3. Internal Tok2Vec: This is the standard TextCat config afaik. I wanted to use the pre-trained Tok2Vec from nb_core_news_lg in the hopes that its tokenizer might have some transfer value to my task, since the dataset for my use case is limited.

4. Unfrozen Tok2Vec: Exactly so, which is why I froze it, to avoid training the Tok2Vec. Again, I was under the impression that "disabled" was the config argument that stopped pipes from running during training.

That's very interesting!

This is the way I see it:

Have a pre-trained pipeline that is very good at understanding (tokenizing, annotating) a language.
Add a module to the end of the pipeline that uses all (or as much as relevant) data in the doc for text classification.

The way I see it, this approach should allow for better information being fed into the relevant TextCat module.
As far as I can tell however, the Tok2Vec component is a central part of the current standard TextCat module learning.
That seems like adding unnecessary work and making the work less precise with my current information.
The way I see it now, having a CNN or other ensemble able to take as input a full pre-trained language pipeline output should yield maximally dense and precise information to the TextCat model, improving predictions.
Theoretically, of course. I may be misunderstanding the function of the Tok2Vec module.
However, I do like the idea of having "a single source of truth" with a common Tok2Vec module for all pipelines.

**Reply by adrianeboyd** [reply]
Sorry this is confusing! The docs are currently slightly incorrect about what happens with frozen components. They do not run during training but do run during evaluation on the dev corpus. In v3.0, it only makes sense to either have tok2vec + all listening components frozen or all of them non-frozen. Any mixture of frozen+non-frozen is going to lead to errors or poor performance.

The change coming in v3.1 is to allow any component to "annotate" during training (#7767), so a frozen component wouldn't be updated, but it could still run so that later components can see its predictions. This allows the `tok2vec` to run and provide `doc.tensor` for `textcat`, but not be modified.

The `tok2vec` in `nl_core_news_lg` is probably not that useful for `textcat` because it's been trained for tagging+parsing tasks. (This is what my quick test with the textcat project demos with `en_core_web_sm` showed, in any case. It's possible models with vectors would be different, or there are cases where it works better. I haven't really experimented enough to know more.)

A shared `tok2vec` component is going to work best if you have a training corpus that includes annotation for all the listening components, so if you had tags+parses+cats on your training data. Even then, for really different kinds of tasks, the performance could be worse than with separate `tok2vec` components.

In pretrained pipelines like `nl_core_news_lg`, the syntax-y components (`tagger`, `morphologizer`, `parser`) are trained on the same corpus with a shared `tok2vec` and the `ner` component, which is often trained on a different corpus, has a separate internal `tok2vec`. See a bit more info about the pipeline design here: https://spacy.io/models#design-cnn

In spacy v2, all components had their own separate internal `tok2vec` layers, so it's not an unusual choice for `textcat` to have its own internal `tok2vec`.

Disabled components are not run at all during training or evaluation. In the pretrained pipelines we train the `senter` at the same time as the syntax components (with its own teeny internal `tok2vec`) and then disable it right before the pipeline is packaged. This allows us to ship the `senter` model with the package, but if you load the default pipeline, `senter` doesn't run at all and `parser` does sentence segmentation.

**Reply by Economax** [reply]
Aha, so disable and frozen are different in terms of evaluation, not training?

v3.1 sounds like it will open for what I'm trying to do, which is nice! When is v3.1 expected to be out?

I see, so the argument is that a different token to vector mapping may be preferable depending on the classification task?
E.g. one Tok2Vec for each pipeline?
Though in that case, why bother having a common Tok2Vec pipeline for some components?
Wouldn't the argument be that there is some transfer learning or generalization going on with a shared Tok2Vec?

Also, I don't really see why we would need annotation for all the listening components when training the TextCat.
What I want is for a pre-trained pipeline to provide _useful information_ to a TextCat pipeline.
The pre-trained model is great for parsing norwegian. I just want to take all the data generated from the standard pipeline, input it in a TextCat model, and train it based on whether it managed to predict the correct label.

Maybe we're thinking of different cases here? I want to maximize the quality and quantity of information passed to my TextCat pipeline, not just give it the text contents of a Doc or Example. 

I.e. I want to use the information generated by all of the pipelines in the pretrained model to help categorize my text.

**Reply by adrianeboyd** [reply]
* `disable` is a **pipeline** setting that determines whether the component runs at all (training/evaluation/anywhere)
* `frozen_components` is a **training** setting that determines whether the component is updated during training

Some combinations of components benefit more from joint training than others. As a really simple example: knowing that a particular token is parsed as `adj` could be used to improve the tag, but probably isn't going to help you predict the cats.

With spacy v3.0, I think your best option is to experiment with a standalone `textcat` component with an internal `tok2vec`.

We don't have an estimate for the release date for v3.1 at this point, sorry.

**Reply by Economax** [reply]
Perfect, that's a very clear way of stating it!

Sure, though in my estimation, that information could very well help prediction. Hard to say a priori of course!

I'll figure something out, the standard textcat did a decent job at SA!

I might give a custom TextCat pipeline a go and see if the additional info from the pre-trained model can help.
Maybe I'm way off base, but at least I'll know!

Thanks for your help so far, I'll come back to this thread with test results if I can remember it.

**Reply by Economax** [reply]
One thing: Can I pre-train the internal Tok2Vec in the TextCat with the corpora from nb_core_news_lg?
